runtime/onert/backend/train/TensorBuilder.cc

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "TensorBuilder.h"
+
+#include "Tensor.h"
+
+namespace onert
+{
+namespace backend
+{
+namespace train
+{
+
+TensorBuilder::TensorBuilder(const std::shared_ptr<TensorRegistry> &tensor_reg,
+                             const std::string planner_id)
+  : _tensor_reg{tensor_reg}, _tensor_mgr{new TensorManager(tensor_reg, planner_id)}
+{
+  /* empty */
+}
+
+void TensorBuilder::registerTensorInfo(const ir::OperandIndex &index, const ir::OperandInfo &info,
+                                       ir::Layout layout)
+{
+  _tensor_info_map.emplace(index, info);
+  _as_constants[index] = info.isConstant();
+
+  // Train backend supports only one layout as NHWC
+  assert(layout == ir::Layout::NHWC);
+  assert(!info.isDynamic());
+
+  // NOTE For now, whether or not to build operands to trainable tensor depends on whether
+  //      the corresponding operand is constant.
+  if (_as_constants[index])
+  {
+    auto tensor = std::make_unique<TrainableTensor>(info, layout);
+    _tensor_reg->setTrainableTensor(index, std::move(tensor));
+  }
+  else
+  {
+    auto tensor = std::make_unique<Tensor>(info, layout);
+    _tensor_reg->setNonConstTensor(index, std::move(tensor));
+  }
+}
+
+void TensorBuilder::registerBackwardTensorInfo(const ir::OperandIndex &index,
+                                               const ir::OperandInfo &info, ir::Layout layout)
+{
+  _backward_tensor_info_map.emplace(index, info);
+
+  // Train backend supports only one layout as NHWC
+  assert(layout == ir::Layout::NHWC);
+  assert(!info.isDynamic());
+
+  // NOTE For now, whether or not to build operands to trainable tensor depends on whether
+  //      the corresponding operand is constant.
+  assert(_as_constants[index] == info.isConstant());
+  if (_as_constants[index])
+  {
+    auto tensor = std::make_unique<GradientTensor>(info, layout);
+    _tensor_reg->setGradientTensor(index, std::move(tensor));
+  }
+  else
+  {
+    auto tensor = std::make_unique<DerivativeTensor>(info, layout);
+    _tensor_reg->setDerivativeTensor(index, std::move(tensor));
+  }
+}
+
+void TensorBuilder::notifyFirstUse(const ir::OperandIndex &index)
+{
+  // TODO Support momory plan
+  if (_as_constants[index])
+  {
+    _tensor_mgr->claimTrainablePlan(index);
+  }
+  else
+  {
+    _tensor_mgr->claimNonConstPlan(index);
+  }
+}
+
+void TensorBuilder::notifyLastUse(const ir::OperandIndex &)
+{
+  // TODO Support momory plan
+}
+
+void TensorBuilder::notifyBackwardFirstUse(const ir::OperandIndex &index)
+{
+  // TODO Support momory plan
+  if (_as_constants[index])
+  {
+    _tensor_mgr->claimGradientPlan(index);
+  }
+  else
+  {
+    _tensor_mgr->claimDerivativePlan(index);
+  }
```

Comment:
What is the meaning of `_as_constants`?
Is it related to differentiability?

Reply:
> What is the meaning of _as_constants?

`_as_constants` contains whether each operand is a constant. Since there is no way to distinguish trainable operands(like weights) yet, `_as_constants` is used to distinguish trainable operands.

> Is it related to differentiability?

No, rather than that,  `_as_constants` was used to distinguish between derivative tensors that will be backpropagated and gradient tensors that will be used to update trainable tensors here.
Also we may need derivative tensors of the corresponding trainable tensors. If so, we have to modify this change like this:
```suggestion
  if (_as_constants[index])
  {
    _tensor_mgr->claimGradientPlan(index);
  }
  _tensor_mgr->claimDerivativePlan(index);
```
But I'm not sure if those derivative tensors is necessary.